### PR TITLE
Refactor materials using Label and Density type aliases

### LINKED ISSUE #219: Define `Union{String,Symbol}` as a `const` for labels

### PR DESCRIPTION
**Pull request by mforets**
Closes https://github.com/ONSAS/ONSAS.jl/issues/219

### CODE REVIEW

File: src/Elements/Tetrahedron.jl [hunk 1]
```diff
@@ -23,23 +24,22 @@ See [[Belytschko]](@ref) and [[Gurtin]](@ref) for more details.
...
 struct Tetrahedron{dim,T<:Real,N<:AbstractNode{dim,T}} <: AbstractElement{dim,T}
     nodes::SVector{4,N}
-    label::Symbol
+    label::Label
     function Tetrahedron(nodes::SVector{4,N},
-                         label=:no_labelled_element) where
+                         label::Label=NO_LABEL) where
              {dim,T<:Real,N<:AbstractNode{dim,T}}
         @assert dim == 3 "Nodes of a tetrahedron element must be 3D."
         return new{dim,T,N}(nodes, Symbol(label))
```
Comment (mvanzulli) [inline src/Elements/Tetrahedron.jl]:
```suggestion
        new{dim,T,N}(nodes, Symbol(label))
```

File: src/Elements/Tetrahedron.jl [hunk 2]
```diff
@@ -23,23 +24,22 @@ See [[Belytschko]](@ref) and [[Gurtin]](@ref) for more details.
...
     end
 end
 
-function Tetrahedron(n₁::N, n₂::N, n₃::N, n₄::N, label=:no_labelled_element) where {N<:AbstractNode}
-    return Tetrahedron(SVector(n₁, n₂, n₃, n₄), Symbol(label))
+function Tetrahedron(n₁::N, n₂::N, n₃::N, n₄::N, label::Label=NO_LABEL) where {N<:AbstractNode}
+    return Tetrahedron(SVector(n₁, n₂, n₃, n₄), label)
```
Comment (mvanzulli) [inline src/Elements/Tetrahedron.jl]:
```suggestion
    Tetrahedron(SVector(n₁, n₂, n₃, n₄), label)
```

File: src/Elements/Tetrahedron.jl [hunk 3]
```diff
@@ -23,23 +24,22 @@ See [[Belytschko]](@ref) and [[Gurtin]](@ref) for more details.
...
 end
 
 "Constructor for a `Tetrahedron` element without nodes and a `label`. This function is used to create meshes via GMSH."
-function Tetrahedron(label::L=:no_labelled_face) where {L<:Union{String,Symbol}}
-    return Tetrahedron(SVector(Node(0, 0, 0), Node(0, 0, 0), Node(0, 0, 0), Node(0, 0, 0)),
-                       Symbol(label))
+function Tetrahedron(label::Label=NO_LABEL)
+    return Tetrahedron(SVector(Node(0, 0, 0), Node(0, 0, 0), Node(0, 0, 0), Node(0, 0, 0)), label)
```
Comment (mvanzulli) [inline src/Elements/Tetrahedron.jl]:
```suggestion
    Tetrahedron(SVector(Node(0, 0, 0), Node(0, 0, 0), Node(0, 0, 0), Node(0, 0, 0)), label)
```

File: src/Elements/TriangularFace.jl [hunk 4]
```diff
@@ -14,9 +15,9 @@ A `TriangularFace` represents an element composed by three `Node`s.
...
 struct TriangularFace{dim,T<:Real,N<:AbstractNode{dim,T}} <: AbstractFace{dim,T}
     nodes::SVector{3,N}
-    label::Symbol
+    label::Label
     function TriangularFace(nodes::SVector{3,N},
-                            label=:no_labelled_face) where
+                            label::Label=NO_LABEL) where
              {dim,T<:Real,N<:AbstractNode{dim,T}}
         @assert 2 ≤ dim ≤ 3 "TriangularFace is only defined for 2 < dim ≤ 3"
         return new{dim,T,N}(nodes, Symbol(label))
```
Comment (mvanzulli) [inline src/Elements/TriangularFace.jl]:
```suggestion
        new{dim,T,N}(nodes, Symbol(label))
```

File: src/Elements/TriangularFace.jl [hunk 5]
```diff
@@ -25,14 +26,14 @@ end
 
 "Constructor for a `TriangularFace` element considering the nodes `n₁` `n₂` and `n₃`."
 function TriangularFace(n₁::N, n₂::N, n₃::N,
-                        label::L=:no_labelled_face) where
-         {dim,T<:Real,N<:AbstractNode{dim,T},L<:Union{String,Symbol}}
-    return TriangularFace(SVector(n₁, n₂, n₃), Symbol(label))
+                        label::Label=NO_LABEL) where
+         {dim,T<:Real,N<:AbstractNode{dim,T}}
+    return TriangularFace(SVector(n₁, n₂, n₃), label)
```
Comment (mvanzulli) [inline src/Elements/TriangularFace.jl]:
```suggestion
    TriangularFace(SVector(n₁, n₂, n₃), label)
```

File: src/Elements/TriangularFace.jl [hunk 6]
```diff
@@ -25,14 +26,14 @@ end
...
 end
 
 "Constructor for a `TriangularFace` element without nodes and a `label`. This function is used to create meshes via GMSH."
-function TriangularFace(label::L=:no_labelled_face) where {L<:Union{String,Symbol}}
-    return TriangularFace(SVector(Node(0, 0), Node(0, 0), Node(0, 0)), Symbol(label))
+function TriangularFace(label::Label=NO_LABEL)
+    return TriangularFace(SVector(Node(0, 0), Node(0, 0), Node(0, 0)), label)
```
Comment (mvanzulli) [inline src/Elements/TriangularFace.jl]:
```suggestion
    TriangularFace(SVector(Node(0, 0), Node(0, 0), Node(0, 0)), label)
```

File: src/Elements/Truss.jl [hunk 7]
```diff
@@ -23,9 +24,9 @@ See [[ANLE]](@ref).
...
     nodes::SVector{2,N}
     cross_section::G
-    label::Symbol
+    label::Label
     function Truss(nodes::SVector{2,N}, g::G,
-                   label=:no_labelled_element) where
+                   label::Label=NO_LABEL) where
              {dim,T<:Real,N<:AbstractNode{dim,T},G<:AbstractCrossSection}
         @assert 1 ≤ dim ≤ 3 "Nodes of a truss element must comply  1 < dim < 3 ."
         return new{dim,T,N,G}(nodes, g, Symbol(label))
```
Comment (mvanzulli) [inline src/Elements/Truss.jl]:
```suggestion
        new{dim,T,N,G}(nodes, g, Symbol(label))
```

File: src/Elements/Truss.jl [hunk 8]
```diff
@@ -34,14 +35,14 @@ end
 
 "Constructor for a `Truss` element considering the nodes `n₁` and `n₂` and the cross-section `g`."
 function Truss(n₁::N, n₂::N, g::G,
-               label::L=:no_labelled_face) where
-         {dim,T<:Real,N<:AbstractNode{dim,T},G<:AbstractCrossSection,L<:Union{String,Symbol}}
-    return Truss(SVector(n₁, n₂), g, Symbol(label))
+               label::Label=NO_LABEL) where
+         {dim,T<:Real,N<:AbstractNode{dim,T},G<:AbstractCrossSection}
+    return Truss(SVector(n₁, n₂), g, label)
```
Comment (mvanzulli) [inline src/Elements/Truss.jl]:
```suggestion
    Truss(SVector(n₁, n₂), g, label)
```

File: src/Materials/HyperElastic.jl [hunk 9]
```diff
@@ -1,38 +1,34 @@
...
-                          label::L=:no_labelled_mat) where
-             {T<:Real,F<:Function,R<:Union{Nothing,Real},L<:Union{Symbol,String}}
-        return new{T,F,R}(params, Ψ, ρ, Symbol(label))
+    "Density (`nothing` for static cases)."
+    ρ::Density
+    "Material label."
+    label::Label
+    function HyperElastic(params::Vector{T}, Ψ::F, ρ::Density,
+                          label::Label=NO_LABEL) where {T<:Real,F<:Function}
+        return new{T,F}(params, Ψ, ρ, Symbol(label))
```
Comment (mvanzulli) [inline src/Materials/HyperElastic.jl]:
```suggestion
        new{T,F}(params, Ψ, ρ, Symbol(label))
```

File: src/Materials/IsotropicLinearElastic.jl [hunk 10]
```diff
@@ -1,48 +1,43 @@
...
-                                    label::L=:no_labelled_mat) where
-             {ET<:Real,NT<:Real,RT<:Union{Nothing,Real},L<:Union{Symbol,String}}
-        return new{ET,NT,RT}(E, ν, ρ, Symbol(label))
+    "Density (`nothing` for static cases)."
+    ρ::Density
+    "Material label."
+    label::Label
+    function IsotropicLinearElastic(E::ET, ν::NT, ρ::Density,
+                                    label::Label=NO_LABEL) where {ET<:Real,NT<:Real}
+        return new{ET,NT}(E, ν, ρ, Symbol(label))
```
Comment (mvanzulli) [inline src/Materials/IsotropicLinearElastic.jl]:
```suggestion
        new{ET,NT}(E, ν, ρ, Symbol(label))
```

File: src/Materials/IsotropicLinearElastic.jl [hunk 11]
```diff
@@ -1,48 +1,43 @@
...
     end
 end
-"Material `IsotropicLinearElastic` constructor with no density parameter `ρ`."
-function IsotropicLinearElastic(E::Real, ν::Real,
-                                label::L=:no_labelled_mat) where {L<:Union{Symbol,String}}
+function IsotropicLinearElastic(E::ET, ν::NT, label::Label=NO_LABEL) where {ET<:Real,NT<:Real}
     return IsotropicLinearElastic(E, ν, nothing, label)
```
Comment (mvanzulli) [inline src/Materials/IsotropicLinearElastic.jl]:
```suggestion
    IsotropicLinearElastic(E, ν, nothing, label)
```

File: src/Materials/IsotropicLinearElastic.jl [hunk 12]
```diff
@@ -1,48 +1,43 @@
...
-"Material `IsotropicLinearElastic` constructor with Lamé parameters `λ`, `G` and density `ρ`. 
-See [this ref](https://en.wikipedia.org/wiki/Lam%C3%A9_parameters)."
-function IsotropicLinearElastic(; λ::Real, G::Real, ρ::R=nothing,
-                                label::L=:no_labelled_mat) where
-         {R<:Union{Nothing,Real},L<:Union{Symbol,String}}
+function IsotropicLinearElastic(; λ::Real, G::Real, ρ::Density=nothing, label::Label=NO_LABEL)
     E = G * (3λ + 2G) / (λ + G)
     ν = λ / (2 * (λ + G))
-
     return IsotropicLinearElastic(E, ν, ρ, Symbol(label))
```
Comment (mvanzulli) [inline src/Materials/IsotropicLinearElastic.jl]:
```suggestion
    IsotropicLinearElastic(E, ν, ρ, Symbol(label))
```

File: src/Materials/NeoHookean.jl [hunk 13]
```diff
@@ -1,62 +1,55 @@
...
+    "Material label."
+    label::Label
+    function NeoHookean(K::T, G::T, ρ::Density, label::Label=NO_LABEL) where {T<:Real}
         if ρ isa Real
             ρ > 0 || error("Density must be positive.")
         end
         @assert K ≥ 0 "The bulk modulus `K` must be positive."
         @assert G ≥ 0 "The shear modulus or second Lamé parameter `μ` must be positive."
-        return new{T,R}(K, G, ρ, Symbol(label))
+        return new{T}(K, G, ρ, Symbol(label))
```
Comment (mvanzulli) [inline src/Materials/NeoHookean.jl]:
```suggestion
        new{T}(K, G, ρ, Symbol(label))
```

File: src/Materials/NeoHookean.jl [hunk 14]
```diff
@@ -1,62 +1,55 @@
...
-
-    # Compute λ, μ and K (μ = G) given E and ν
+function NeoHookean(; E::Real, ν::Real, ρ::Density=nothing, label::Label=NO_LABEL)
+    # Compute λ, μ and K (μ = G) given E and ν.
     λ = E * ν / ((1 + ν) * (1 - 2 * ν))
     G = E / (2 * (1 + ν))
     K = λ + 2 * G / 3
-
-    return NeoHookean(K, G, ρ, Symbol(label))
+    return NeoHookean(K, G, ρ, label)
```
Comment (mvanzulli) [inline src/Materials/NeoHookean.jl]:
```suggestion
    NeoHookean(K, G, ρ, label)
```

File: src/Materials/SVK.jl [hunk 15]
```diff
@@ -1,60 +1,52 @@
...
+    "Material label."
+    label::Label
+    function SVK(λ::T, G::T, ρ::Density, label::Label=NO_LABEL) where {T<:Real}
         if ρ isa Real
             ρ > 0 || error("Density must be positive.")
         end
         @assert λ ≥ 0 "The first Lamé parameter `λ` must be positive."
         @assert G ≥ 0 "The second Lamé parameter or shear modulus `G` must be positive."
-        return new{T,R}(λ, G, ρ, Symbol(label))
+        return new{T}(λ, G, ρ, Symbol(label))
```
Comment (mvanzulli) [inline src/Materials/SVK.jl]:
```suggestion
        new{T}(λ, G, ρ, Symbol(label))
```

File: src/Materials/SVK.jl [hunk 16]
```diff
@@ -1,60 +1,52 @@
...
-         {R<:Union{Nothing,Real},L<:Union{Symbol,String}}
-
-    # Compute λ and μ (μ = G) given E and ν
+function SVK(; E::Real, ν::Real, ρ::Density=nothing, label::Label=NO_LABEL)
+    # Compute λ and μ (μ = G) given E and ν.
     λ = E * ν / ((1 + ν) * (1 - 2 * ν))
     G = E / (2 * (1 + ν))
-
-    return SVK(λ, G, ρ, Symbol(label))
+    return SVK(λ, G, ρ, label)
```
Comment (mvanzulli) [inline src/Materials/SVK.jl]:
```suggestion
    SVK(λ, G, ρ, label)
```

**Review by mvanzulli — CHANGES_REQUESTED** [review]
Maybe we can delete `return` from these files

**Comment by mvanzulli** [comment]
Can we add `Nothing` into `Label` const type ? This will save memory when creating a lot of elements with gmsh.

File: src/Elements/Tetrahedron.jl (hunk 1 above)
Comment (mforets) [reply src/Elements/Tetrahedron.jl]: But then why not re-enable[`always_use_return`](https://github.com/ONSAS/ONSAS.jl/blob/main/.JuliaFormatter.toml#L2) again?

EDIT: Ah, now i see you meant *delete* :)

**Comment by mforets** [comment]
Maybe as a separate step? Note that the inner constructors still convert back to `Symbol`. `NO_LABEL` might be turned to `nothing` with that change too.

**Comment by mvanzulli** [comment]
> Maybe as a separate step? Note that the inner constructors still convert back to `Symbol`. `NO_LABEL` might be turned to `nothing` with that change too.

Okay I will add an issue for the next step

**Review by mvanzulli — APPROVED** [review]

**Pull request merged**